src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+extern crate byteorder;
+
+use super::{Counter, Histogram};
+use std::io::{self, Cursor, Read, Write};
+use std;
+use super::num::ToPrimitive;
+use self::byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+#[path = "tests.rs"]
+#[cfg(test)]
+mod tests;
+
+#[path = "benchmarks.rs"]
+#[cfg(all(test, feature = "bench_private"))]
+mod benchmarks;
+
+const V2_COOKIE_BASE: u32 = 0x1c849303;
+
+const V2_COOKIE: u32 = V2_COOKIE_BASE | 0x10;
+
+const V2_HEADER_SIZE: usize = 40;
+
+/// Errors that occur during serialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SerializeError {
+    /// A count above i64::max_value() cannot be zig-zag encoded, and therefore cannot be
+    /// serialized.
+    CountNotSerializable,
+    /// Internal calculations cannot be represented in `usize`. Use smaller histograms or beefier
+    /// hardware.
+    UsizeTypeTooSmall,
+    /// An i/o operation failed.
+    IoError
+}
+
+impl std::convert::From<std::io::Error> for SerializeError {
+    fn from(_: std::io::Error) -> Self {
+        SerializeError::IoError
+    }
+}
+
+/// Serializer for the V2 binary format.
+/// Serializers are intended to be re-used for many histograms.
```

Comment:
So the use-case is if you want to serialize or deserialize *lots* of histograms right after each other? That doesn't seem all that common? And also comes with an ergonomic penalty.. One alternative would be to instead have `serialize` and `serialize_with`, where the latter takes an `&mut Vec`. This would *allow* users in that particular use-case to have efficient repeated serialization, without incurring the usability penalty for those who just want a simple `to_bytes()` serialization.